Enhancement: option to send HTTP over unix socket

ae-govau opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------

Add one method to send HTTP traffic (for any backend) over a unix socket.

I'm not sure this is the best way, but it certainly seemed an easy way without a large refactor.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

Closes #4287 

Checklist
---------

- [X] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [X] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [X] I have run `gofmt` on the code in all commits.
- [X] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [X] I'm done! This pull request is ready for review.

Happy to add tests but wanted feedback on approach first. 

While adding syntax like `-r rest:unix:/tmp/socket` looked simple in the original issue, the way the code is structured it's difficult to get access to the `http.Transport` object from within the `rest` backend. Hence I added a global option, which means this would also work for tunnelling other HTTP based backends over a socket if desired.

Our intended use-case is in conjunction with the `rest-server` (see https://github.com/restic/rest-server/pull/272) where a central backup server essentially makes the following SSH calls out to hosts to back them up:

```bash
rest-server --listen unix:/tmp/socket --data /path/to/data &
scp restic host:/tmp
ssh -R /tmp/socket:/tmp/socket host /tmp/restic -use-unix-socket-for-http /tmp/socket -r rest:http://ignored/repo
```

(the `ignored` bit is ugly)

We can do the above with ephermal ports today, but unix sockets feels like it might be cleaner for this-use case.

MichaelEischer reviewed (changes requested):
I strongly dislike adding a CLI option that then goes on to override the path passed to the backend. That just feels wrong. (It would also introduce multiple weird failure cases for other backends).

My suggestion is to use `"github.com/peterbourgon/unixtransport"` instead. With a call to `unixtransport.Register(tr)` it  becomes possible to use `rest:http+unix:///path/to/socket` without requiring a messy CLI option. That adds a new (very small) dependency, but that's better than the alternative.

Commit:
Enhancement: option to send HTTP over unix socket

add tests for unix socket connection

switch HTTP rest-server test to use any free port

allow rest-server test graceful shutdown opportunity

ae-govau commented:
> My suggestion is to use `"github.com/peterbourgon/unixtransport"` instead. With a call to `unixtransport.Register(tr)` it becomes possible to use `rest:http+unix:///path/to/socket` without requiring a messy CLI option. That adds a new (very small) dependency, but that's better than the alternative.

Thanks, that library looks good and that's a much cleaner outcome. Have made change as suggested. On my system I also needed to make a fix to that library (so that it will ignore HTTP proxies when invoked). We might want to wait to verify that patch is accepted (and version bumped) before merging this one:
https://github.com/peterbourgon/unixtransport/pull/7

I've tested the above against the `rest-server` PR I made here:
https://github.com/restic/rest-server/pull/272

and it seems to work.

MichaelEischer commented on doc/030_preparing_a_new_repo.rst:
```diff
@@ -201,15 +201,16 @@ scheme like this:
     $ restic -r rest:https://host:8000/ init
     $ restic -r rest:https://user:pass@host:8000/ init
     $ restic -r rest:https://user:pass@host:8000/my_backup_repo/ init
+    $ restic -r rest:http+unix:///tmp/rest.socket:/my_backup_repo/ init
```
I wondering whether  we should also add `https+unix` here? Although I have to admit that I have no clue how the HTTPS certificate and host name checks would behave. So it might even be a good idea not to mention `https+unix`, hmm.

MichaelEischer reviewed (commented):
The code is ok now, but I'd really like to see a small test, that connecting to a `http+unix` server does work. Otherwise, there's a too large chance that we accidentally break this feature.

Waiting for a fix to the proxy support is probably a good idea. And https://github.com/restic/rest-server/pull/272 also has to be completed first.

ae-govau commented:
Happy to add tests once the server-part lands.

ae-govau commented:
Integration test added and cleaned up regular HTTP based `rest-server` test at same time.

ae-govau commented:
Rebased to fix `go.sum` conflicts. This remains ready for review.

MichaelEischer committed:
CI: uses rest-server from master branch until unix sockets are released

MichaelEischer committed:
add changelog for rest unix socket support

MichaelEischer reviewed (approved):
LGTM. I've fixed the test failures (we'll have to switch rest-server back to `latest` once the next version is released) and there's now a changelog. Please take a look at it, afterwards we should be ready to merge this PR.

ae-govau commented:
Looks good. Thanks for fixing up and following this through.
